### PR TITLE
feat: log node ids with smart client

### PR DESCRIPTION
**Pull request by AmeanAsad**
## Changes:
- When making requests directly to nodes, we now have the ability to log their ids even if the initial fetch fails without response.

### CODE REVIEW

File: src/client.js [hunk 1]
```diff
@@ -83,17 +85,16 @@ export class Saturn {
       }
     }
 
-    let origins = options.origins
-    if (!origins || origins.length === 0) {
-      const replacementUrl = options.url ?? options.cdnURL
-      origins = [replacementUrl]
+    let nodes = options.nodes
+    if (!nodes || nodes.length === 0) {
+      const replacementUrl = options.node ?? { url: this.opts.cdnURL }
```
Comment (guanzo) [inline src/client.js:90]:
```suggestion
      const replacementNode = options.node ?? { url: this.opts.cdnURL }
```

File: src/client.js [hunk 2]
```diff
@@ -103,11 +104,11 @@ export class Saturn {
...
         clearTimeout(connectTimeout)
-        return { res, url, controller }
+        return { res, url, node, controller }
       } catch (err) {
-        throw new Error(
-          `Non OK response received: ${res.status} ${res.statusText}`
-        )
+        const error = err
+        error.node = node
+        throw new Error(error)
```
Comment (guanzo) [inline src/client.js:111]: I'm not sure what happens if you pass an object to the `Error` constructor, it expects a string.
```suggestion
        err.node = node
        throw err
```

File: src/client.js [hunk 3]
```diff
@@ -294,10 +303,10 @@ export class Saturn {
         return
       }
       if (opts.raceNodes) {
-        const origins = nodes.slice(i, i + Saturn.defaultRaceCount).map((node) => node.url)
-        opts.origins = origins
+        const origins = nodes.slice(i, i + Saturn.defaultRaceCount)
+        opts.nodes = origins
```
Comment (guanzo) [inline src/client.js:307]:
```suggestion
        opts.nodes = nodes.slice(i, i + Saturn.defaultRaceCount)
```

File: src/client.js (hunk 1 above)
Comment (AmeanAsad) [reply src/client.js]: fixed this, had to change the following line as well

File: src/client.js [hunk 4]
```diff
@@ -119,17 +119,18 @@ export class Saturn {
...
 
-    let res, url, controller
+    let res, url, controller, node
     try {
-      ({ res, url, controller } = await fetchPromises)
+      ({ res, url, controller, node } = await fetchPromises)
 
       abortRemainingFetches(controller, controllers)
+      log.nodeId = node.id
       log = Object.assign(log, this._generateLog(res, log), { url })
```
Comment (guanzo) [inline src/client.js:134]: the assignment on line `133` will be overwritten by `this._generateLog(res, log)`

perhaps this for line 231?

`log.nodeId = log.nodeId ?? headers.get('saturn-node-id')`

Comment (AmeanAsad) [reply src/client.js:134]: agreed, changed

**Pull request merged**